src/courses/mappers/04.md

```diff
@@ -0,0 +1,342 @@
+author: Charles Hu
+---
+
+## What Is a Mapper?
+
+A mapper is a design pattern used to correlate (or map) items in two different objects with one another. Mappers are useful in that they allow us to connect items in objects that are nominally different but semantically the same. A result of this is that we can easily transform one object type into another through the use of mappings which define the direct relationship of items in each object type.
+
+Here are some scenarios which demonstrate some key aspects of mappers:
+
+::: details Transferring IDs
+Say we have to transfer the credentials of an employee who is moving from Business A to Business B. The data formats the businesses use for IDs are as follows:
+```json
+// Business A
+{Name, DoB, Age, Title}
+
+// Business B
+{employee, employeeBirthday, employeeAge, jobTitle}
+```
+
+How do we transfer John's credentials from Business A to Business B?
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+```
+
+What we can do is create a mapping which correlates the items from Business A's ID scheme to Business B's:
+```json
+{employee: Name, employeeBirthday: DoB, employeeAge: Age, jobTitle: Title}
+```
+
+With this, we can then develop a mapper which takes John's credentials from Business A and transforms it to Business B's format as so:
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+
+ ||
+\  /
+ \/
+
+MAPPER
+
+ ||
+\  /
+ \/
+
+{employee: 'John Doe', employeeBirthday: 10-6-1992, employeeAge: 32, jobTitle: 'Security Technician'}
+``` 
+
+The important thing to note here is that mappers rely on underlying mappings which match semantically similar fields between two objects. These matches allow us to correctly convert each item in one object to the other.
+:::
+
+::: details Translating a foreign language
+Say we are given the job to translate the traditional Chinese sentence '我愛貓' into English.
+
+Without knowledge of the language, this is an impossible task.
+
+However, say we now also had a mapper which utilized the following mapping:
+```json
+{'愛': 'love', '我': 'I', '貓': 'cats'}
+```
+
+Now this task becomes much more easier. We can apply this mapper which will translate the sentence from '我愛貓' to 'I love cats' and vice versa.
+
+The important thing to note here is that mappers make translating between two objects trivial, which means that anyone can do it and we can automate mappers to perform object conversions for us.
+:::
+
+
+## What Is an OHDF Mapper?
+
+An OHDF mapper is a mapper that allows the underlying conversion infrastructure found in OHDF Converters to correlate certain objects or values from one overarching security tool data format to another overarching security tool data format. In the case of the SAF tool suite, these mappers allow for the conversion of any given data format to HDF (\*-to-HDF) and vice versa (HDF-to-\*) using the tools provided by the existing conversion infrastructure.
```

Comment:
what are the tools provided by the existing conversion infrastructure?

Reply:
Addressed.